sdk/eventhub/azure-eventhubs-aio-partitionmanager-blobstorage/azure/eventhub/aio/partitionmanager/blobstorage/blobstoragepm.py

```diff
@@ -0,0 +1,125 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+from typing import Iterable, Dict, Any
+import logging
+from collections import defaultdict
+import asyncio
+from azure.eventhub.eventprocessor import PartitionManager, OwnershipLostError
+from azure.core.exceptions import ResourceModifiedError, ResourceExistsError, AzureError
+from azure.storage.blob.aio import ContainerClient
+
+logger = logging.getLogger(__name__)
+UPLOAD_DATA = ""
+
+
+class BlobPartitionManager(PartitionManager):
+    """An PartitionManager that uses Azure Blob Storage to store the partition ownership and checkpoint data.
+
+    This class implements methods list_ownership, claim_ownership, and update_checkpoint that are defined in class
+    azure.eventhub.eventprocessor.PartitionManager of package azure-eventhub.
+
+    """
+    def __init__(self, container_client: ContainerClient):
+        """
+
+        :param container_client: The Azure Blob Storage Container client.
+        """
+        self._container_client = container_client
+        self._cached_ownership_dict = defaultdict(dict)  # type: Dict[str, Dict[str, Any]]
+        # lock each partition for list_ownership, claim_ownership and update_checkpoint etag doesn't get out of sync
+        # when the three methods are running concurrently
+        self._cached_ownership_locks = defaultdict(asyncio.Lock)
+
+    async def list_ownership(self, eventhub_name: str, consumer_group_name: str) -> Iterable[Dict[str, Any]]:
+        try:
+            blobs = self._container_client.list_blobs(include=['metadata'])
+        except AzureError as azure_err:  # list_blobs has exhausted retry
+            logger.exception("An exception occurred during list_ownership for eventhub %r consumer group %r."
+                             " An empty ownership list is returned",
+                             eventhub_name, consumer_group_name, exc_info=azure_err)
+            return []
```

Comment:
Won't the partition manager try to take ownership of all partitions if you return an empty list?

Reply:
EventProcess only claims ownership on the returned list. Empty list means no ownership.

Reply:
The previous version of the [code](https://github.com/Azure/azure-sdk-for-python/blob/787d65fb21f06ddb529d936f836656da0db9830b/sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py#L119) tries to take ownership of all partitions not currently owned.

Reply:
Previous version doesn't have multiple EventProcessors so that worked. 
EventProcessor now has load balancing so the code has changed a lot.

Reply:
I mixed list_ownership and claim_ownership by mistake. This should raise the exception instead of return []